### ISSUE TITLE
Remove partial methods still influencing nullability flow analysis?

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#42671]
**Version Used**: 
C# Tools   3.6.0-2.20167.1+926207070b0c62d6a76aec4627ec26dee17870bb

**Steps to Reproduce**:
```C#
#nullable enable
using System.Diagnostics.CodeAnalysis;

partial class Program
{
    static void Main()
    {
        string? value = null;
        Write(ref value);
        string notnull = value;
    }

    static partial void Write([NotNull] ref string? s);
}
```

**Expected Behavior**:
There is no implementation of the partial method `Write`, so the call site will be removed, and the `[NotNull]` that's on the `Write` argument won't apply, so this should warn on the `string notnull = value;` assignment.

**Actual Behavior**:
No warning.

**Comment by sharwell** [comment]
I actually think this should warn on the `Write` method definition due to the failure to adhere to the stated post-condition.

**Comment by YairHalberstadt** [comment]
I don't know if it makes sense for partial/conditional methods to state post conditions, in the same way as they must be void returning and not declare out parameters.

**Comment by jcouv** [comment]
In addition to Sam's and Yair's proposals (warn on un-implemented partial, or warn on post-condition attribute on partial method, respectively), I want to point out that we already tolerate and evaluate annotations on methods that get omitted, ie. that don't technically honor their contract (`Debug.Assert` for example).

Overall, this is a fairly narrow problem, which seem only applicable to `NotNull` and `MemberNotNull`. I think that other attributes (`[MaybeNull/When]`, `[NotNullWhen]`, ...) don't pose a problem.
So I'd prioritize this fairly low. I'm okay with status quo, but if we can get a quick design decision I'm okay to change too :-)

**Comment by stephentoub** [comment]
> I want to point out that we already tolerate and evaluate annotations on methods that get omitted, ie. that don't technically honor their contract

Yes, but this is different.  Debug.Assert isn't making the condition true, it's just saying it should be true.  This partial method case is actually removing the thing that's promising to make it true.

**Comment by jcouv** [comment]
> Debug.Assert isn't making the condition true,

I don't quite see it that way. It says "when I return, X will be not-null". You could do that with an assignment to X, but you could also do that with a `throw` :-)
```void M([NotNull] out string? p) => throw null!;` is okay and honors the contract of `[NotNull]` and that of `out`.

**Comment by stephentoub** [comment]
Debug.Assert is not expected to change the value of the thing it's asserting about.  This partial example is.

Maybe we'll need to agree to disagree. :)
